rewrite-java-21/src/main/java/org/openrewrite/java/isolated/ReloadableJava21ParserVisitor.java

```diff
@@ -779,12 +785,24 @@ public J visitInstanceOf(InstanceOfTree node, Space fmt) {
         return new J.InstanceOf(randomId(), fmt, Markers.EMPTY,
                 convert(node.getExpression(), t -> sourceBefore("instanceof")),
                 convert(node.getType()),
-                node.getPattern() instanceof JCBindingPattern b ?
-                        new J.Identifier(randomId(), sourceBefore(b.getVariable().getName().toString()), Markers.EMPTY, emptyList(), b.getVariable().getName().toString(),
-                                type, typeMapping.variableType(b.var.sym)) : null,
+                getNodePattern(node.getPattern(), type),
                 type);
     }
 
+    private @Nullable J getNodePattern(PatternTree pattern, JavaType type){
+        if (pattern instanceof  JCBindingPattern b) {
+            return new J.Identifier(randomId(), sourceBefore(b.getVariable().getName().toString()), Markers.EMPTY, emptyList(), b.getVariable().getName().toString(),
+                    type, typeMapping.variableType(b.var.sym));
+        } else if (pattern instanceof JCRecordPattern r) {
+            int saveCursor = cursor;
+            int endCursor = max(endPos(r), cursor);
+            cursor = endCursor;
+            return new J.Unknown(randomId(), whitespace(), Markers.EMPTY, new J.Unknown.Source(randomId(), whitespace(), Markers.EMPTY, source.substring(saveCursor, endCursor)));
+        } else {
+            return null;
+        }
```

Comment:
I can't find it now anymore, but I think I added a comment somewhere that we might be able to merge these two branches into one and only return `null` if the `pattern` is `null`.

What I find curious is that the `ParenthesizedPatternTree` only exists in the Java 17 compiler. Does that mean that parenthesized guards are not allowed in Java 21? Was that maybe only ever allowed in previews of this feature?

It seems like Java 21 instead has the `AnyPatternTree`: `if (r instanceof R(_)) {}`. See also https://docs.oracle.com/javase/specs/jls/se22/html/jls-14.html#jls-14.30.1

Reply:
https://github.com/openrewrite/rewrite/pull/4900#discussion_r1914858017